File: server/shared/schemas/agent.py
from datetime import datetime
from typing import Literal

from typing import Any

from pydantic import BaseModel, Field, field_validator

from shared.schemas.skills import AgentSkillCatalogOut


class AgentCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)
    config: dict | None = None
    system_prompt: str | None = Field(None, max_length=10000)
    model: str | None = Field(None, max_length=100)
    max_steps: int = Field(10, gt=0, le=100)
    tags: list[str] | None = Field(None, max_length=20)
    workspace_id: str


class AgentUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)
    config: dict | None = None
    system_prompt: str | None = Field(None, max_length=10000)
    model: str | None = Field(None, max_length=100)
    max_steps: int | None = Field(None, gt=0, le=100)
    tags: list[str] | None = None


class AgentOut(BaseModel):
    id: str
    name: str
    description: str | None = None
    version: str
    config: dict
    system_prompt: str | None = None
    model: str | None = None
    max_steps: int
    tags: list
    workspace_id: str
    created_by: str
    created_at: datetime

    model_config = {"from_attributes": True}


class AgentDetailOut(AgentOut):
    """`GET /agents/{id}` only — includes discoverable skills for the editor UI."""

    skill_catalog: AgentSkillCatalogOut


class AgentListOut(BaseModel):
    agents: list[AgentOut]
    total: int
    page: int
    page_size: int


class RunAttachmentIn(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    mime_type: str | None = Field(None, max_length=120)
    text: str | None = Field(None, max_length=120_000)


class RunCreate(BaseModel):
    goal: str = Field(min_length=1, max_length=8000)
    attachments: list[RunAttachmentIn] | None = Field(None, max_length=12)
    model: str | None = Field(
        None,
        max_length=100,
        description="If set, overrides Agent.model for this run only (OpenAI model id or vLLM served name).",
    )
    replace_run_id: str | None = Field(
        None,
        max_length=21,
        description="If set, reset this existing run and re-queue it instead of creating a new row.",
    )
    continue_run_id: str | None = Field(
        None,
        max_length=21,
        description="If set, re-queue this run with a new user turn (multi-turn in one run row).",
    )


class RunConversationPatch(BaseModel):
    """Persisted UI thread: legacy flat list, or v2 object with `linear` (worker) + `tree` (branches)."""

    conversation: list[dict] | dict

    @field_validator("conversation")
    @classmethod
    def _conversation_size(cls, v):
        if isinstance(v, list):
            if len(v) > 500:
                raise ValueError("conversation list too long")
            return v
        if isinstance(v, dict):
            lin = v.get("linear")
            if isinstance(lin, list) and len(lin) > 400:
                raise ValueError("conversation.linear too long")
            tree = v.get("tree")
            if isinstance(tree, dict):
                msgs = tree.get("messages")
                if isinstance(msgs, list) and len(msgs) > 600:
                    raise ValueError("conversation.tree.messages too large")
        return v


class RunPinBody(BaseModel):
    pinned: bool


class RunFeedbackBody(BaseModel):
    feedback: Literal["positive", "negative"] | None = None


class RunOut(BaseModel):
    id: str
    agent_id: str
    status: str
    goal: str | None = None
    steps: list
    output: dict | None = None
    error: str | None = None
    total_steps: int
    triggered_by: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    pinned_at: datetime | None = None
    feedback: str | None = None
    conversation: Any | None = None

    model_config = {"from_attributes": True}


class RunListOut(BaseModel):
    runs: list[RunOut]
    total: int
    page: int
    page_size: int

File: src/services/agent.ts
import { api } from "./api";

const PREFIX = "/agents";

export interface AgentOut {
  id: string;
  name: string;
  description: string | null;
  version: string;
  config: Record<string, unknown>;
  system_prompt: string | null;
  model: string | null;
  max_steps: number;
  tags: unknown[];
  workspace_id: string;
  created_by: string;
  created_at: string;
}

export interface AgentListOut {
  agents: AgentOut[];
  total: number;
  page: number;
  page_size: number;
}

export interface AgentCreateInput {
  name: string;
  description?: string | null;
  workspace_id: string;
  config?: Record<string, unknown> | null;
  system_prompt?: string | null;
  model?: string | null;
  max_steps?: number;
  tags?: string[] | null;
}

export interface AgentUpdateInput {
  name?: string;
  description?: string | null;
  config?: Record<string, unknown> | null;
  system_prompt?: string | null;
  model?: string | null;
  max_steps?: number;
  tags?: string[] | null;
}

export interface AgentSkillSourceItem {
  id: string;
  label: string;
  sources: ("mozi" | "agents" | "config")[];
}

export interface AgentSkillCatalogOut {
  items: AgentSkillSourceItem[];
  selected: string[];
}

/** `GET /agents/:id` — includes skill list for the editor. */
export interface AgentDetailOut extends AgentOut {
  skill_catalog: AgentSkillCatalogOut;
}

export interface CreateLocalSkillInput {
  name: string;
  title?: string | null;
  description?: string | null;
}

/** Step persisted on `AgentRun` after execution (LangChain tool loop). */
export interface AgentRunStep {
  tool: string;
  input: string;
  output: string;
}

export interface RunOut {
  id: string;
  agent_id: string;
  status: string;
  goal: string | null;
  steps: AgentRunStep[];
  output: Record<string, unknown> | null;
  error: string | null;
  total_steps: number;
  triggered_by: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  pinned_at?: string | null;
  feedback?: string | null;
}

export interface RunListOut {
  runs: RunOut[];
  total: number;
  page: number;
  page_size: number;
}

export interface RunAttachmentIn {
  name: string;
  mime_type?: string | null;
  text?: string | null;
}

export interface RunCreateInput {
  goal: string;
  attachments?: RunAttachmentIn[];
  /** Overrides agent default model for this run (OpenAI id or vLLM served name). */
  model?: string | null;
}

export const agentApi = {
  list(workspaceId: string, page = 1, pageSize = 20) {
    return api.get<AgentListOut>(`${PREFIX}/`, {
      params: { workspace_id: workspaceId, page, page_size: pageSize },
    });
  },

  get(agentId: string) {
    return api.get<AgentDetailOut>(`${PREFIX}/${agentId}`);
  },

  createLocalSkill(agentId: string, body: CreateLocalSkillInput) {
    return api.post<AgentSkillCatalogOut>(`${PREFIX}/${agentId}/skills/create-local`, body);
  },

  importSkillFolder(agentId: string, formData: FormData) {
    return api.postFormData<AgentSkillCatalogOut>(`${PREFIX}/${agentId}/skills/import`, formData);
  },

  importSkillZip(agentId: string, formData: FormData) {
    return api.postFormData<AgentSkillCatalogOut>(
      `${PREFIX}/${agentId}/skills/import-zip`,
      formData,
    );
  },

  create(body: AgentCreateInput) {
    return api.post<AgentOut>(`${PREFIX}/`, body);
  },

  update(agentId: string, body: AgentUpdateInput) {
    return api.patch<AgentOut>(`${PREFIX}/${agentId}`, body);
  },

  startRun(agentId: string, body: RunCreateInput) {
    return api.post<RunOut>(`${PREFIX}/${agentId}/run`, body);
  },

  listRuns(agentId: string, page = 1, pageSize = 30) {
    return api.get<RunListOut>(`${PREFIX}/${agentId}/runs`, {
      params: { page, page_size: pageSize, _: Date.now() },
      cache: "no-store",
    });
  },

  getRun(runId: string) {
    return api.get<RunOut>(`${PREFIX}/runs/${runId}`, {
      params: { _: Date.now() },
      cache: "no-store",
    });
  },

  cancelRun(runId: string) {
    return api.post<void>(`${PREFIX}/runs/${runId}/cancel`, {});
  },

  deleteRun(runId: string) {
    return api.delete<void>(`${PREFIX}/runs/${runId}`);
  },

  pinRun(runId: string, pinned: boolean) {
    return api.post<RunOut>(`${PREFIX}/runs/${runId}/pin`, { pinned });
  },

  setRunFeedback(runId: string, feedback: "positive" | "negative") {
    return api.post<RunOut>(`${PREFIX}/runs/${runId}/feedback`, { feedback });
  },
};
